services/relay.c

```diff
@@ -0,0 +1,66 @@
+#include "jd_services.h"
+#include "interfaces/jd_pins.h"
+#include "jacdac/dist/c/relay.h"
+
+struct srv_state {
+    SRV_COMMON;
+    uint8_t intensity;
+    uint8_t variant;
+    uint32_t max_switching_current;
+    relay_params_t params;
+};
+
+REG_DEFINITION(                                   //
+    relay_regs,                                   //
+    REG_SRV_COMMON,                         //
+    REG_U8(JD_RELAY_REG_CLOSED),                 //
+    REG_OPT8(JD_RELAY_REG_VARIANT),                //
+    REG_OPT32(JD_RELAY_REG_MAX_SWITCHING_CURRENT),            //
+)
+
+void reflect_register_state(srv_t* state) {
+    // active
+    if (state->intensity) {
+        pin_set(state->params.pin_relay_drive, (state->params.drive_active_lo ? 0 : 1));
+        pin_set(state->params.pin_relay_led, (state->params.led_active_lo ? 0 : 1));
+    }
+    // inactive
+    else {
+        pin_set(state->params.pin_relay_drive, (state->params.drive_active_lo ? 1 : 0));
+        pin_set(state->params.pin_relay_led, (state->params.led_active_lo ? 1 : 0));
+    }
+
+    jd_send_event_ext(state, (state->intensity > 0) ? JD_RELAY_EV_ON : JD_RELAY_EV_OFF, 0, 0);
+}
+
+void relay_process(srv_t * state) {
+    (void)state;
+    // bool server_state = state->intensity;
+
+    // if (state->params.pin_relay_feedback != NO_PIN && server_state != (bool)pin_get(state->params.pin_relay_feedback))
```

Comment:
I can remove this once I confirm the purpose of this GPIO dedicated to relay feedback

Reply:
sounds good